### PR TITLE
Add frame bit to encryption method

### PR DESCRIPTION
**Pull request by Trivo25**
This PR introduces a naive `frame bit` to the `encrypt` and `decrypt` method. The last message block is marked by an additional `1`

### CODE REVIEW

File: src/lib/provable/crypto/encryption.ts [hunk 1]
```diff
@@ -68,3 +72,64 @@ function decrypt(
...
+
+  let sponge = new Poseidon.Sponge();
+  sponge.absorb(sharedSecret.x); // don't think we need y, that's enough entropy
+
+  // encryption
+  let cipherText = [];
+  for (let i = 0; i < message.length; i++) {
+    let keyStream = sponge.squeeze();
+    let encryptedChunk = message[i].add(keyStream);
+    if (message.length - 1 === i) encryptedChunk = encryptedChunk.add(1);
```
Comment (Trivo25) [inline src/lib/provable/crypto/encryption.ts:93]: this is the only line added

File: src/lib/provable/crypto/encryption.ts [hunk 2]
```diff
@@ -68,3 +72,64 @@ function decrypt(
...
+  let sponge = new Poseidon.Sponge();
+  sponge.absorb(sharedSecret.x);
+  let authenticationTag = cipherText.pop();
+
+  // decryption
+  let message = [];
+  for (let i = 0; i < cipherText.length; i++) {
+    let keyStream = sponge.squeeze();
+    let messageChunk = cipherText[i].sub(keyStream);
+    if (cipherText.length - 1 === i) messageChunk = messageChunk.sub(1);
```
Comment (Trivo25) [inline src/lib/provable/crypto/encryption.ts:125]: this is the only line added

File: src/lib/provable/crypto/encryption.ts [hunk 3]
```diff
@@ -40,6 +42,38 @@ function encrypt(message: Field[], otherPublicKey: PublicKey) {
...
+
+  let sponge = new Poseidon.Sponge();
+  sponge.absorb(sharedSecret.x); // don't think we need y, that's enough entropy
+
+  // encryption
+  let cipherText = [];
+  for (let i = 0; i < message.length; i++) {
+    let keyStream = sponge.squeeze();
+    let encryptedChunk = message[i].add(keyStream);
+    if (message.length - 1 === i) encryptedChunk = encryptedChunk.add(1);
```
Comment (Trivo25) [inline src/lib/provable/crypto/encryption.ts:61]: this is the only line added

File: src/lib/provable/gadgets/foreign-field.ts [hunk 4]
```diff
@@ -739,8 +739,7 @@ class Sum {
 function assertLessThan(x: Field3, y: bigint | Field3) {
   let y_ = Field3.from(y);
 
-  // constant case
-
+  // constant case, y = constant, x = constant
```
Comment (Trivo25) [inline src/lib/provable/gadgets/foreign-field.ts:742]: this is just a left over recommendation to adjust the comments as well

**Review by mitschabaude — COMMENTED** [review]
What you're doing is a bit different than the recommendation.
I think the goal of the rec is to make the final message chunk unambiguously different from the other ones. To do so, the final chunk is (0, msg[i)) while the other chunks are (1, msg[i]).

I don't think it's a problem to swap 0 and 1 here, but from what I can tell, just adding 1 doesn't achieve the goal of distinguishing the final chunk.

Caveat: I didn't take the time to understand the math of the attack and remediation.

**Comment by mitschabaude** [comment]
To replicate the frame bit exactly, we'd need to completely redo the encryption method since the inputs can't be full field elements; they'd need to be at most 253-bit chunks.

Another option could be to understand the motivation for the frame bit at a fundamental level and find an alternative remediation to the attack which is more circuit-friendly.

**Comment by mitschabaude** [comment]
I'd probably do a new version where the message is required to be `Bytes`, and then accumulate 31 bytes = 248 bits into one chunk. That still leaves room for the frame bit without overflowing the field.

And arguably, bytes is a more natural representation of an arbitrary message than field elements

**Comment by mitschabaude** [comment]
> I'd probably do a new version where the message is required to be `Bytes`, and then accumulate 31 bytes = 248 bits into one chunk.

Or maybe not bytes but something more flexible like `(Bool | UInt8 | UInt32 | UInt64)[]`. that makes it easier to use (and more efficient) with large elements like individual uint64s. And the encryption method packs as many inputs as possible into 1 chunk such that there is still room for a frame bit.

**Comment by Trivo25** [comment]
https://github.com/o1-labs/o1js/pull/1729

**Pull request closed without merging**